### PR TITLE
Go to definition functionality added to hotkey and context menu

### PR DESCRIPTION
**Pull request by robmck1995**
Fixes https://github.com/marimo-team/marimo/issues/1126

For this one, I set the hotkey to Mod+Shift+, but let me know if there's a better one.

I'm unsure if this is structured correctly, specifically, the utility functions specified in `cell-editor.tsx`. Let me know if there is a better way to structure it. Basically, I wasn't sure how to ensure we could use the same functionality with the right-click menu and the hotkeys. The solution works, but there is a bit of code repetition.

Created this with the help of [Glide](https://glide.agenticlabs.com)! The change in this document only covers adding the hotkey, I manually added the right-click after as it was a small change:
https://glide.agenticlabs.com/sharing?userId=65bae76b7fbfb1c20ac1f3b8&taskId=ghM1mqX

### CODE REVIEW

File: frontend/src/components/editor/cell/cell-context-menu.tsx [hunk 1]
```diff
@@ -81,6 +87,21 @@ export const CellActionsContextMenu = ({ children, ...props }: Props) => {
...
+    {
+      label: "Go to Definition",
+      icon: <SearchIcon size={13} strokeWidth={1.5} />,
+      handle: () => {
+        const { getEditorView } = props;
+        const editorView = getEditorView();
+        if (editorView) {
+          const variableName = getWordUnderCursor(editorView?.state);
+          const focusCellId = getCellIdOfDefinition(variables, variableName);
+          if (focusCellId) {
```
[the remark this answers is not in the document]
Comment (robmck1995) [reply frontend/src/components/editor/cell/cell-context-menu.tsx:99]: I think that `getWordUnderCursor` should always return some string. And `getCellIdofDefinition` will only return a value if the variable is defined in the variables array so I think this should not cause errors, but not 100% sure about this

File: frontend/src/components/editor/cell/code/cell-editor.tsx [hunk 2]
```diff
@@ -394,3 +418,29 @@
...
+export const getCellIdOfDefinition = (
+  variables: Variables,
+  variableName: string,
+) => {
+  const variable = variables[variableName as VariableName];
+  if (!variable || variable.declaredBy.length === 0) {
+    return null;
+  }
+  const focusCellId = variable.declaredBy[0];
+  return focusCellId;
```
[the remark this answers is not in the document]
Comment (mscolnick) [reply frontend/src/components/editor/cell/code/cell-editor.tsx:461]: yea this is ok - you can define something twice, but its an error and going to the first one will be ok

File: frontend/src/core/hotkeys/hotkeys.ts [hunk 3]
```diff
@@ -258,6 +258,11 @@ const DEFAULT_HOT_KEY = {
     group: "Navigation",
     key: "Mod-Shift-s",
   },
+  "cell.goToDefinition": {
+    name: "Go to Definition",
+    group: "Navigation",
+    key: "Mod-Shift-,",
+  },
```
[the remark this answers is not in the document]
Comment (mscolnick) [reply frontend/src/core/hotkeys/hotkeys.ts:308]: this is fine

File: frontend/src/core/cells/cells.ts [hunk 4]
```diff
@@ -231,6 +231,25 @@ const {
       scrollKey: null,
     };
   },
+  focusCellAtDefinition: (
```
Comment (mscolnick) [inline frontend/src/core/cells/cells.ts:234]: Even though its kinda similar to `focusCell`, I think we can make this a function outside of the reducer in `core/codemirror/go-to-definition`. You can access the store by `store.get(atomName)`.

File: frontend/src/components/editor/cell/code/cell-editor.tsx [hunk 5]
```diff
@@ -410,3 +434,29 @@ const CellCodeMirrorEditor = React.forwardRef(
 CellCodeMirrorEditor.displayName = "CellCodeMirrorEditor";
 
 export const CellEditor = memo(CellEditorInternal);
+
+export const getWordUnderCursor = (state: EditorState) => {
```
Comment (mscolnick) [inline frontend/src/components/editor/cell/code/cell-editor.tsx:438]: i think we can move all this logic to another file `core/codemirror/go-to-definition.ts`. 
we should be able to just expose one function `goToDefinition(view: EditorView)`

File: frontend/src/components/editor/cell/cell-context-menu.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (mscolnick) [reply frontend/src/components/editor/cell/cell-context-menu.tsx:99]: i can play around with some cases - vscode does always show you `Go To Definition`, but does no-op when you don't have a work under it

File: frontend/src/components/editor/cell/code/cell-editor.tsx (hunk 5 above)
Comment (robmck1995) [reply frontend/src/components/editor/cell/code/cell-editor.tsx:438]: Perfect

File: frontend/src/core/codemirror/cells/extensions.ts [hunk 6]
```diff
@@ -52,6 +52,7 @@ export function cellMovementBundle(
     toggleHideCode,
     aiCellCompletion,
   } = callbacks;
+  const variables = useVariables()
```
Comment (mscolnick) [inline frontend/src/core/codemirror/cells/extensions.ts:55]: you can use `store.get(variables)` and can likely push this into `goToDefinition`

Comment (robmck1995) [reply frontend/src/core/codemirror/cells/extensions.ts:55]: Will add, thanks for the tip

**Review by mscolnick — APPROVED** [review]
nice contrib!

**Pull request merged**